### PR TITLE
Speedups for precommit hooks

### PR DESCRIPTION
**Pull request by eparis**
This (along with the separate munger change) greatly reduces the pre-commit hook time. From like 30+ seconds down to 8.5 seconds to build everything, 2.5 seconds  to get swagger (that stinks) and another 2 seconds to do everything else in the hook.

Notice this does make it easier for the gendeepcopy and genconversion to be wrong when run by hand. This is the exact same as gen\* can be wrong when run by hand if you didn't build new copies.  The git pre-commit hook builds everything, so it should always be right (as would travis et al once you open a PR)

### CODE REVIEW

**Comment by mikedanese** [comment]
@thockin

**Comment by zmerlynn** [comment]
So this now even more assumes that your local build is up to date, it looks like? i.e. if you haven't done a `make` in a little while, your precommits might just fail? (which sucks for some workflows, like `git bisect`)

File: hack/lib/util.sh [hunk 1]
```diff
@@ -171,7 +171,11 @@ kube::util::gen-doc() {
   done <"${dest}/.files_generated"
 
   # put the new generated file into the destination
-  find "${tmpdir}" -exec rsync -pt {} "${dest}" \; >/dev/null
+  files=($(find "${tmpdir}" -type f))
+  if [[ "${#files[@]}" -ne 0 ]]; then
+    cp -f "${files[@]}" "${dest}"
```
Comment (thockin) [inline hack/lib/util.sh]: I don't think this does the right thing.  Consider:

```
thockin@freakshow:/tmp$ rm -rf a z

thockin@freakshow:/tmp$ mkdir -p a/b/c/d

thockin@freakshow:/tmp$ touch a/b/c/d/file-in-a-b-c-d

thockin@freakshow:/tmp$ touch a/b/c/file-in-a-b-c

thockin@freakshow:/tmp$ mkdir -p z/b/c/d

thockin@freakshow:/tmp$ find a -type f
a/b/c/d/file-in-a-b-c-d
a/b/c/file-in-a-b-c

thockin@freakshow:/tmp$ find z -type f

thockin@freakshow:/tmp$ f=($(find a -type f))

thockin@freakshow:/tmp$ cp -f "${f[@]}" z

thockin@freakshow:/tmp$ find z -type f
z/file-in-a-b-c
z/file-in-a-b-c-d
```

File: hooks/pre-commit [hunk 2]
```diff
@@ -41,31 +41,13 @@ fi
...
-fi
-
-# Check for py files without the required boilerplate.
-files=($(git diff --cached --name-only --diff-filter ACM | grep "\.py" | grep -v -e "third_party" -e "Godeps"))
-if [[ ${#files} -gt 0 ]]; then
-  files_need_boilerplate+=($("${boiler}" "py" "${files[@]}"))
-fi
-
-if [[ "${#files_need_boilerplate[@]}" -ne 0 ]]; then
+out=($(hack/verify-boilerplate.sh))
```
Comment (thockin) [inline hooks/pre-commit]: Won't this catch files that may have snuck in but are not part of this commit?  Maybe not a huge deal, but we could pass "${files[@]}" to the script and have it use the commandline if present, else find files itself

File: hack/update-generated-conversions.sh [hunk 3]
```diff
@@ -44,7 +56,7 @@ import (
 // AUTO-GENERATED FUNCTIONS START HERE
 EOF
 
-	GOPATH=$(godep path):$GOPATH go run cmd/genconversion/conversion.go -v ${version} -f - >>  $TMPFILE
```
Comment (thockin) [inline hack/update-generated-conversions.sh]: There's a reason we force preconditions by 'go run' - we have been bitten over and over by stale builds of tools "passing" precommit, only to fail CI (or worse, be committed).

I think the right pattern to follow is something like:

update-generated-conversions.sh:
```
#!/bin/bash
make WHAT=cmd/genconversion
exec $(realpath $(dirname $0))/update-generated-conversions__prebuilt.sh
```

update-generated-conversions__prebuilt.sh
```
#!/bin/bash
# test for existence of the binary/binaries we need
# run them
```

Then have precommit run a `make` followed by all of the __prebuilt scripts. Humans run the regular scripts.

**Comment by thockin** [comment]
Eric,

Are you still working on this.  It's driving me batty a little more each day...

**Comment by eparis** [comment]
i keep hoping to get back to it every afternoon.  today looks ok.

File: hooks/pre-commit (hunk 2 above)
Comment (eparis) [inline hooks/pre-commit]: It will. I've got some other branch to make that possible, but for now, just doing it for the whole repo is really fast.  I can pull this out if you want...

**Comment by eparis** [comment]
hopefully other comments have been addressed.  PTAL.

**Comment by thockin** [comment]
LGTM

**Comment by eparis** [comment]
@k8s-bot test this

**Comment by cjcullen** [comment]
@eparis Do you want to squash commits before I merge?

**Comment by eparis** [comment]
squashed as seemed appropriate...  each commit should be its own thing now.

**Comment by cjcullen** [comment]
Thanks!
